### PR TITLE
go_bears_on_bsd.md: Add Go on BSD project

### PR DESCRIPTION
**Pull request by yzgyyang**
**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: _projects/go_bears_on_bsd.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+`dependency_management` provides Python APIs for dependency management on various languages and different systems. It is used extensively by coala, however, it lacks a lot of the key implementation and documentation in its current state.
+
+The student will implement features to coala’s automated management system. `dependency_management` should be able to manage language runtimes and operate language-specific package managers for dependencies. It should also resolve package names and versions automatically on multiple operating systems.
+
+The end goal is to be able to bootstrap and recursively install all dependencies on any systems without anything installed other than a base system. The scope of a complete solution is too large. The student made a proposal with suitable scope, which is to automate the installation of a Go Linter Bear onto bare-metal BSDs. The project may also provide a blueprint for future implementation and enhancement.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: `The student` isnt suitable for a project intro.

Just describe your project here.

For example, what are all the current go linters in coala-bears?

And what are the typical problems with Go dependencies?  Especially how does Go do versioning?

e.g. https://github.com/coala/coala-bears/issues/2457
See also https://github.com/coala/coala-bears/issues/1472 and https://github.com/coala/coala-bears/issues/1488 and other problems we have encountered.
And https://github.com/coala/coala/issues/2752#issuecomment-382411052 for Go problems which are even older and probably never solved and likely are now unsolvable.

As you can see, Go is ok with breaking old tools, and de-supports old versions quite quickly.

In the milestones `The student ...` below is correct.

File: _projects/go_bears_on_bsd.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+##### PREPARATION/BONDING
+
+* The student has identified most of the related problems.
+* The student has came up with a workflow of how dependencies would be managed across multiple operating systems and package managers.
+
+##### CODING PHASE 1
+
+* The dependencies of the Go Linter are identified and compiled.
+* Support for BSD is implemented in `dependency_management`.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: I guess you are only using mocks for test coverage here?

File: _projects/go_bears_on_bsd.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+##### CODING PHASE 2
+
+* The student has integrated a CI server to the working repo for testing on FreeBSD.
+* At least part of the scope has been implemented and tested.
+* The student has wrote sufficent tests and documentation for the same.
+
+##### CODING PHASE 3
+
+* The proposed scope is completly implemented and tested.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: itemise what you are going to achieve.

File: _projects/go_bears_on_bsd.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+##### CODING PHASE 1
+
+* The dependencies of the Go Linter are identified and compiled.
+* Support for BSD is implemented in `dependency_management`.
+* The student has started implementing bootstrap features for bare-metal systems.
+
+##### CODING PHASE 2
+
+* The student has integrated a CI server to the working repo for testing on FreeBSD.
+* At least part of the scope has been implemented and tested.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: which part.  Be precise.  This is your work plan.  Your milestone tasks will be based on these.

File: _projects/go_bears_on_bsd.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+* The student has wrote sufficent tests and documentation for the same.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: tests and documentation should be separate items.

for tests, you should be having >100% test coverage

File: _projects/go_bears_on_bsd.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
+---
+collaborating_projects:
+  - coala
+desc: "Explore improvements to coala's dependency management through having Go Linter installed in *BSD system automatially."
+difficulty: medium
+initiatives:
+  - GSoC
+  - RGSoC
```
Comment (yukiisbored) [inline _projects/go_bears_on_bsd.md]: This project isn't part of RGSoC afaik.

File: _projects/go_bears_on_bsd.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  - "At least one patch to the coala core should be accepted and merged."
+  - "The proposal needs to show familiarity with coala's dependency management implementation."
+status: in_progress
+tags:
+  - Automation
+  - Installation
+  - Dependency
+---
+
+`coala-bears` are of two types, Native Bears which implement all the business logic themselves and Linter Bears which wrap existing linters and checkers. These Linter Bears have dependencies which need to be installed before they can be executed.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: Native Bears can also depend on anything as well, including python packages (e.g. nltk) or command line tools (e.g. git)

**Comment by yukiisbored** [comment]
I think the main issue with this proposal is lack of information. Try being a bit more specific and lay out the details :+1:

File: _projects/go_bears_on_bsd.md (hunk 2 above)
Comment (yzgyyang) [reply _projects/go_bears_on_bsd.md]: I think so for now, yeah.

**Comment by yzgyyang** [comment]
@gitmate-bot why still 81f1927 😢

**Comment by yukiisbored** [comment]
@gitmate-bot rebase

**Comment by yukiisbored** [comment]
ack https://github.com/coala/projects/pull/577/commits/882f0ce53f2d2005c85aafdaf5873a9ce0560629

**Comment by yukiisbored** [comment]
@gitmate-bot ff

**Comment by yukiisbored** [comment]
@jayvdb Help :/

File: _projects/go_bears_on_bsd.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+`dependency_management` aims to provide Python APIs for dependency management on various languages and different systems. It is used extensively by coala, however, it lacks most of the key implementation and documentation in its current state.
+
+coala now has [7 bears](https://github.com/coala/coala-bears/blob/master/bears/go/) for the Go language. They need specific versions of Go and some externally supplied packages, and currently Go linter bears just assume they are all in place and work. Go language dependencies are currently managed unversioned ("If you're using an externally supplied package and worry that it might change in unexpected ways, the simplest solution is to copy it to your local repository. This is the approach Google takes internally", see [Go Official FAQ](https://golang.org/doc/faq#get_version)).
+
+The end goal is to be able to bootstrap and recursively install all dependencies on any systems without anything installed other than a base system, but the scope of a complete solution is too large. The student made a proposal with suitable scope, which is to automate the installation of Go linter bears onto bare-metal BSDs.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: "all dependencies" or "all go dependencies "?

Also breaking lines at 80 cpl helps reviewing

File: _projects/go_bears_on_bsd.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+##### CODING PHASE 1
+
+* The student has identified and compiled the dependencies of the Go Linter into `coala-bears`.
+* The student has deployed a personal FreeBSD CI server and linked it to the working repo.
+* The student has implemented support for FreeBSD's default package manager [pkg](https://github.com/freebsd/pkg) and test mocks in `dependency_management`.
+* The student has started implementing bootstrap features for bare-metal systems.
+
+##### CODING PHASE 2
+
+* The student has ensured Repology metadata for the Go Linter is in place and correct.
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: I think phase one should have at least one link to freebsd via repology completed, or at least investigated. So we learn about problems early, and any change in milestones occurs at end of phase 1

Then phase two finishes the process.

File: _projects/go_bears_on_bsd.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+executed.
+
+`dependency_management` aims to provide Python APIs for dependency management on
+various languages and different systems. It is used extensively by coala,
+however, it lacks most of the key implementation and documentation in its
+current state.
+
+coala now has [7
+bears](https://github.com/coala/coala-bears/blob/master/bears/go/) for the Go
+language. They need specific versions of Go and some externally supplied
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: we also have bears written in go which validate other file types.  at least the toml linter.  These should also be included here

File: _projects/go_bears_on_bsd.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+packages, and currently Go linter bears just assume they are all in place and
+work. Go language dependencies are currently managed unversioned ("If you're
+using an externally supplied package and worry that it might change in
+unexpected ways, the simplest solution is to copy it to your local repository.
+This is the approach Google takes internally", see [Go Official
+FAQ](https://golang.org/doc/faq#get_version)).
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: while this is the Google official position, I understand there are/were other commonly used mechanisms for managing to install specific versions of go repos.  branches was one of them.

This doesnt need to be addressed in the first iteration of this project spec, but I believe it should be covered in detail in a spec or in our documentation before this project attempts installing various versions.

File: _projects/go_bears_on_bsd.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+The end goal is to be able to bootstrap and recursively install all dependencies
+on any systems without anything installed other than a base system, but the
+scope of a complete solution is too large. The student made a proposal with
```
Comment (jayvdb) [inline _projects/go_bears_on_bsd.md]: I have said in earlier reviews to remove `The student ..` from the project description.

That type of phrasing belongs in the milestone work plan below.

Comment (yzgyyang) [reply _projects/go_bears_on_bsd.md]: Sorry, missed this one!

**Comment by yukiisbored** [comment]
@gitmate-bot rebase

**Comment by yukiisbored** [comment]
ack https://github.com/coala/projects/pull/577/commits/5d69e44f9bb53876bd944aa570dcef270395c37f

**Comment by yukiisbored** [comment]
@gitmate-bot ff

**Comment by yukiisbored** [comment]
Sheesh

**Comment by yukiisbored** [comment]
unack https://github.com/coala/projects/pull/577/commits/5d69e44f9bb53876bd944aa570dcef270395c37f

**Comment by yukiisbored** [comment]
@xferra Please review ASAP.

File: _projects/go_bears_on_bsd.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
+---
+collaborating_projects:
+  - coala
+desc: "Explore improvements to coala's dependency management through having Go Linter installed in *BSD system automatially."
```
Comment (li-boxuan) [inline _projects/go_bears_on_bsd.md]: spelling error

Comment (yzgyyang) [reply _projects/go_bears_on_bsd.md]: Thanks!

File: _projects/go_bears_on_bsd.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+The end goal is to be able to bootstrap and recursively install all dependencies
+on any systems without anything installed other than a base system, but the
+scope of a complete solution is too large. A proposal has been made with
+suitable scope, which is to automate the installation of Go linter bears onto
+bare-metal BSDs.
+
+Features will mainly be implemented to coala's `dependency_management`. It
+should be able to operate system and language-specific package managers to
+handle dependencies. It should also resolve package names and versions
+automatically using Repology on multiple operating systems.
```
Comment (xferra) [inline _projects/go_bears_on_bsd.md]: Do you plan to use Repology as a service or cache it's data somehow? It's it possible that some package names will not be resolved?

Comment (yzgyyang) [reply _projects/go_bears_on_bsd.md]: I believe this partially lies in @nemaniarjun 's project scope which is to implement Repology as a service. What I'd like to do is to make use of the infrastructure provided by that project and specifically implement and ensure the dependencies required for my project has correct data from Repology.

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Pull request merged**